salesforce/bulkwrite.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	jobId, ok := jobCreateRes["id"] //nolint:varnamelen
+	if !ok {
+		return nil, fmt.Errorf("%w for key %s in %v", ErrKeyNotFound, "id", jobCreateRes)
+	}
+
+	jobIdString, ok := jobId.(string) //nolint:varnamelen
+	if !ok {
+		return nil, fmt.Errorf("%w. expected id to be string, gott %T", ErrInvalidType, jobId)
+	}
+
+	// upload csv and there is no response body other than status code
+	_, err = c.uploadCSV(ctx, jobIdString, config)
+	if err != nil {
+		return nil, fmt.Errorf("uploadCSV failed: %w", err)
+	}
+
+	//
+	data, err := c.completeUpload(ctx, jobIdString)
+	if err != nil {
+		return nil, fmt.Errorf("completeUpload failed: %w", err)
+	}
+
+	id, ok := data["id"].(string) //nolint:varnamelen
+	if !ok {
+		return nil, fmt.Errorf(
+			"%w. expected salesforce job id to be string in response, got %T",
+			ErrInvalidType,
+			data["id"],
+		)
+	}
+
+	completeState, ok := data["state"].(string) //nolint:varnamelen
+	if !ok {
+		return nil, fmt.Errorf(
+			"%w. expected salesforce job state to be string in response, got %T",
+			ErrInvalidType,
+			data["state"],
+		)
+	}
+
+	return &common.BulkWriteResult{
+		JobId: id,
+		State: completeState,
+	}, nil
+}
+
+func joinURLPath(baseURL string, paths ...string) (string, error) {
+	location, err := url.JoinPath(baseURL, paths...)
+	if err != nil {
+		return "", errors.Join(err, common.ErrInvalidPathJoin)
+	}
+
+	return location, nil
+}
+
+func (c *Connector) createJob(ctx context.Context, config common.BulkWriteParams) (*ajson.Node, error) {
+	location, err := joinURLPath(c.BaseURL, "jobs/ingest")
+	if err != nil {
+		return nil, err
+	}
+
+	body := map[string]interface{}{
+		"object":              config.ObjectName,
+		"externalIdFieldName": config.ExternalId,
+		"contentType":         "CSV",
+		"operation":           "upsert",
+		"lineEnding":          "LF",
+	}
+
+	return c.post(ctx, location, body)
+}
+
+func (c *Connector) uploadCSV(ctx context.Context, jobId string, config common.BulkWriteParams) ([]byte, error) {
+	file, err := os.ReadFile(config.FilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload CSV for %s: %w", config.FilePath, errors.Join(err, common.ErrReadFile))
+	}
+
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/batches", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	return c.putCSV(ctx, location, file)
+}
+
+func (c *Connector) completeUpload(ctx context.Context, jobId string) (map[string]interface{}, error) {
+	updateLoadCompleteBody := map[string]interface{}{
+		"state": "UploadComplete",
+	}
+
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := c.patch(ctx, location, updateLoadCompleteBody)
+	if err != nil {
+		return nil, fmt.Errorf("patch failed: %w", errors.Join(err, common.ErrRequestFailed))
+	}
+
+	return ParseNodeToMap(res)
+}
+
+func (c *Connector) GetJobResult(ctx context.Context, jobId string) ([]byte, error) {
```

Comment:
For this function and the other functions below this, I generally wouldn't want our library to export any functions that have `[] byte` or `*ajson.Node` as results since they are not the most user friendly and don't offer any type safety. Eventually these exported methods should all have well-typed results (e.g. a `GetJobResultResponse` struct), since later you'll need to parse the byte array or ajson.Node into structs anyhow. But it's probably hard to test these out right now, so we can go with these for now and then add the types later.